### PR TITLE
nixos/zoneminder: fix some issues with database.createLocally option

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
https://github.com/NixOS/nixpkgs/issues/63152

Not tested, or even built... I haven't used `zoneminder` yet so completely depending on @peterhoeg and @JohnAZoidberg for all testing and review.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by JohnAZoidberg** [comment]
Looks great, thanks!
But I don't quite understand what the problem with `mysql.initialDatabases` is. Can you please explain?

**Comment by aanderse** [comment]
@JohnAZoidberg as per https://nixos.org/nixos/options.html#services.mysql.initialdatabases:
> List of database names and their initial schemas that should be used to create databases on the **first** startup of MySQL.

**Comment by JohnAZoidberg** [comment]
Because you want to create it again if it goes missing after a while?
That makes sense indeed and probably is more useful in many cases than `initialDatabases`.

**Comment by aanderse** [comment]
@JohnAZoidberg I think you're missing the part I emphasized. I emphasized **first**. As in only the **first** time `mysql` starts up. Not the second. Or third. Just the **first**.

For example, if `mysql` is already installed and running on your system, then you install `zoneminder` the database won't be provisioned as desired.

The `initialDatabases` option is moreso for end users I would imagine, and shouldn't really be used in nixos itself that often (if ever). The same is true for the `initialScript` option.

**Comment by JohnAZoidberg** [comment]
Ahh, I do understand what you mean now!

I'd also apply this:
```diff
--- i/nixos/modules/services/misc/zoneminder.nix
+++ w/nixos/modules/services/misc/zoneminder.nix
@@ -147,6 +147,7 @@ in {
           default = "zmuser";
           description = ''
             Username for accessing the database.
+            Not used if <literal>createLocally</literal> is set.
           '';
         };
 
@@ -155,6 +156,7 @@ in {
           default = "zmpass";
           description = ''
             Username for accessing the database.
+            Not used if <literal>createLocally</literal> is set.
           '';
         };
       };
@@ -208,7 +210,7 @@ in {
         package = lib.mkDefault pkgs.mariadb;
         ensureDatabases = [ cfg.database.name ];
         ensureUsers = [{
-          name = cfg.database.username;
+          name = user;
           ensurePermissions = { "${cfg.database.name}.*" = "ALL PRIVILEGES"; };
         }];
       };
```

Because when used locally the Linux user is used for socket authentication as seen in this line:
```
     ZM_DB_USER=${if cfg.database.createLocally then user else cfg.database.username}
```

Also the nginx config validator issues a warning on `location /cache {` that there should be a trailing slash.
You, @aanderse seem to have a lot of experience with nginx. Could adding a slash there have negative consequences?

I build and am running these changes and it looks okay. I have never really fully used this service so I'm not sure if everything is working. I just discovered the issue while trying to make it run for a friend.

**Comment by JohnAZoidberg** [comment]
The service might be broken on unstable but I fixed that in #63151.
To circumvent that you can use `nixpkgs.config.allowBroken = true;` because the package is marked as broken but all the sources are in the cache ;)

**Comment by aanderse** [comment]
> You, @aanderse seem to have a lot of experience with nginx. Could adding a slash there have negative consequences?

@JohnAZoidberg You must be confusing me with someone else. I have never used `nginx` before.

**Comment by peterhoeg** [comment]
> > List of database names and their initial schemas that should be used to create databases on the **first** startup of MySQL.

Nicely caught. Wouldn't it be awesome if the mysql module actually did something similar to what you did manually? Cough, cough...

By the way, I ripped out ZM some time ago, so I'm not able to test it, but maybe a nixos test for this would be good?

**Comment by aanderse** [comment]
> Wouldn't it be awesome if the mysql module actually did something similar to what you did manually? Cough, cough...

That would be great, but possibly not as useful as one might guess. I notice that most applications these days are pretty good at being database aware because they use some form of the migration concept. Often times an application knows whether it has a table structure or not and will create it if not. Exceptions I've worked on recently, where the application is unable to determine whether it should create a database structure or not... `zabbix` and `mediawiki`. I can list about ~6 other services I have worked on recently that **were** database aware enough to create their own table structure.
 
> By the way, I ripped out ZM some time ago, so I'm not able to test it, but maybe a nixos test for this would be good?

Complete hijack to this PR, but what did you replace it with? I have a video camera I'm about to install and looking for opinions :smile: 
@JohnAZoidberg this puts testing solely on you! Comments added as requested.

**Comment by peterhoeg** [comment]
> what did you replace it with?

MotionEye. 

https://gist.github.com/f9000a258afb8b73a6a427475cef9116

*Far* less moving parts and a *way* more sane architecture.

File: nixos/modules/services/misc/zoneminder.nix
Comment (JohnAZoidberg) [inline nixos/modules/services/misc/zoneminder.nix]:
```suggestion
           name = user;
```

Because when used locally the Linux user is used for socket authentication as seen in this line:

```
     ZM_DB_USER=${if cfg.database.createLocally then user else cfg.database.username}
```

**Review by JohnAZoidberg — CHANGES_REQUESTED** [review]

**Review by JohnAZoidberg — CHANGES_REQUESTED** [review]
To make it work you have to add a slash in line 260 to make it look like `location cache/ {`. Otherwise the config validator will complain.
Shouldn't break things, see: https://serverfault.com/questions/607615/using-trailing-slashes-in-nginx-configuration/607731#607731 and https://github.com/yandex/gixy/blob/master/docs/en/plugins/aliastraversal.md

Like I said, I don't have any use for this module either but it doesn't look more broken than before and the changes look very sensible&necessary. => Approve when slash is there.

File: nixos/modules/services/misc/zoneminder.nix
Comment (aanderse) [reply nixos/modules/services/misc/zoneminder.nix]: @JohnAZoidberg I have taken care of this with an assertion instead.

Comment (JohnAZoidberg) [reply nixos/modules/services/misc/zoneminder.nix]: Then you have to revert the comment `Not used if <literal>createLocally</literal> is set.` for the database username.
And I'd also default it to the Linux user if `createLocally` ist set, for conenience.

**Comment by aanderse** [comment]
@peterhoeg I would say at this point this PR more than resolves https://github.com/NixOS/nixpkgs/issues/63152 so if you're happy please feel free to merge :smile:
A quick test might be in order just to make sure the service still boots, but I'll leave that up to you and @JohnAZoidberg to decide.

**Comment by peterhoeg** [comment]
> A quick test might be in order 

You know, if only there was some kind of framework that would allow us to run tests against nixos.... ;-)

**Review by JohnAZoidberg — APPROVED** [review]
Seems to do the same as before, but with easier config.
I ran it and accessed the local webpage before and after.

**Comment by aanderse** [comment]
Thanks @JohnAZoidberg and @peterhoeg!

**Pull request merged**
